### ISSUE TITLE
Hide checkbox for certain rows

### ISSUE THREAD

**Issue by MrMoronIV** [6pac/SlickGrid#343]
I have a working multiselect on my grid, but I want to exclude certain rows from having a checkbox and thus being selectable.

What is the way to go? I have stumbled upon numerous "solutions" from the past but the files they suggest (slick.checkboxselectcolumn.js) are all much smaller than the one included in this package, so I highly doubt overwriting it is a good idea.

What is the the up-to-date solution for this?

**Comment by ghiscoding** [comment]
I don't think that it's possible in the current version as it is. If you have a nice way of doing it, you can contribute to this project with your solution

**Comment by 6pac** [comment]
It would be helpful if you provided links to the best solutions you have found, even if they are not from a compatible fork.

**Comment by MrMoronIV** [comment]
I found this:
https://gist.github.com/jcsjcs/10038299

This one:
https://github.com/mleibman/SlickGrid/blob/000f6de109f4017a51f5f23508f6734a81f39a34/plugins/slick.checkboxselectcolumn.js

And this one:
https://github.com/yakirmanor/SlickGrid/blob/master/plugins/slick.checkboxselectcolumn.js

I'm not sure how these are supposed to be used either, it's quite difficult to understand what's going on exactly an how one should enable the correct options, and where.

**Comment by SatanEnglish** [comment]
@6pac @ghiscoding 
May pay to get it working for detailView at the same time if you end updating this.

Currently trying to work around checkboxes with detailView at the moment.
Will let you know if I come up with anything to help with this while i'm try to deal with combo
Plugin: Row Detail Panel + Plugin: Checkbox

**Comment by SatanEnglish** [comment]
Ok so at first look all you would need to do to incorporate something for this NOT taking to account for the detail expanding rows.

Change 2 functions

Firstly add in a disabled version to formatter

```
        function checkboxSelectionFormatter(row, cell, value, columnDef, dataContext) {
            var UID = createUID() + row;

            if (dataContext) {
                if (dataContext.selectableRow === false) {
                    return _selectedRowsLookup[row]
                        ?"<input id='selector" + UID + "' type='checkbox' checked='checked' disabled><label for='selector" + UID + "'></label>"
                        : "<input id='selector" + UID + "' type='checkbox' disabled><label for='selector" + UID + "'></label>";
                }
                else {
                    return _selectedRowsLookup[row]
                        ? "<input id='selector" + UID + "' type='checkbox' checked='checked'><label for='selector" + UID + "'></label>"
                        : "<input id='selector" + UID + "' type='checkbox'><label for='selector" + UID + "'></label>";
                }
            }
            return null;
        }
```

Secondly add in a check into the select all.

```
        function handleHeaderClick(e, args) {
            if (args.column.id == _options.columnId && $(e.target).is(":checkbox")) {
                // if editing, try to commit
                if (_grid.getEditorLock().isActive() && !_grid.getEditorLock().commitCurrentEdit()) {
                    e.preventDefault();
                    e.stopImmediatePropagation();
                    return;
                }
                
                if ($(e.target).is(":checked")) {
                    var rows = [];
                    for (var i = 0; i < _grid.getDataLength(); i++) {

                        // Get the row and check it's a selectable row before pushing it onto the stack
                        var rowItem = _grid.getDataItem(i);
                        if (rowItem.selectableRow !== false) {
                            rows.push(i);
                        }
                    }
                    _grid.setSelectedRows(rows);
                } else {
                    _grid.setSelectedRows([]);
                }
                e.stopPropagation();
                e.stopImmediatePropagation();
            }
        }
```

**Comment by ghiscoding** [comment]
@MrMoronIV 
First of all, I should have replied why our version of `slick.checkboxselectcolumn.js`, it got bigger when I added support to change styling of the checkbox (see [demo](http://6pac.github.io/SlickGrid/examples/example-checkbox-row-select.html) with Font Awesome)) and then support of show/hide "Select All" button and also toggle it's location filterRow/columnHeaderRow (see [demo](http://6pac.github.io/SlickGrid/examples/example-checkbox-header-row.html)) so that is for why it's much bigger now. 

As for the code you found 
> I found this:
1- https://gist.github.com/jcsjcs/10038299
This one:
2-https://github.com/mleibman/SlickGrid/blob/000f6de109f4017a51f5f23508f6734a81f39a34/plugins/slick.checkboxselectcolumn.js
And this one:
3- https://github.com/yakirmanor/SlickGrid/blob/master/plugins/slick.checkboxselectcolumn.js
I'm not sure how these are supposed to be used either, it's quite difficult to understand what's going on exactly and how one should enable the correct options, and where.

Looking at all 3, they use different terms
1.  uses the term `not_multiselectable` on the `dataContext` which means directly on the item object... already I don't like this one, user should never have to change his object to control the grid. 
2. uses `disabled` row(s), hmmm I'm not too sure this is good either. Disabling an entire row might mean different things, can we export that row or not? can we sort a disabled row? can we filter a disabled row? etc...
3.  uses the term `selectable` on the column definition... now that is better, however how are we suppose to  dynamically change the row from being selectable to not selectable?

How do you envision to say, these rows are selectable but these other rows are not selectable? Do you know this upfront or is it by some properties within your object (meaning dynamically or programmatically calculated)? @SatanEnglish before we implement any code, we need to be able to answer these questions first...

**Comment by SatanEnglish** [comment]
@ghiscoding 
I've gone with 3 for now while I try and work this out for my needs.
This seems the sensible choice as how to determine if you **Can** or **Cannot** check the checkbox will come from the row data given to slickgrid.
This is also where I'd expect to find it again if reordering.

Think if we made the item it looks for to be something less likely to be used in a normal row of data it should be fine. e.g. something like "selectable_row".

JUST ready your idea so (The above could be the easy out of the box setup)

I do like the idea of ALSO adding a way to describe a function based setup.
Main issue I think you will run into is that will deal with the formatter but in case of click / select all you have to grab the datacontext by hand.

Maybe simple canRowBeSelected(function(rowData)

**Comment by ghiscoding** [comment]
@SatanEnglish 
I'm not sure when you read my previous post, but I updated it a few times. I also like number 3 and I think it's the way to go but only internally to the plugin, as for the user then providing a method user can override is ideal and easy for them. As for the arguments, I think it make sense to pass all the same arguments as the Formatter, the more, the better, in many case just passing the `rowData` won't be enough, just like in some Formatters I use only `dataContext` but on some other Formatter I use just the `columnDef`

**Comment by SatanEnglish** [comment]
Others are not there on the click evens is my only concern so while you can use them to show a disabled checkbox you cannot use them for checking the clicks

**Comment by SatanEnglish** [comment]
@ghiscoding  `handleSelectedRowsChanged(e, args)` Is the most concern
args{ rows:[] } <= we only have row none of the cell, value, columnDef (Maybe these are easy to lookup I don't know)

**Comment by SatanEnglish** [comment]
Either way I feel i'm going to need to find a way to have a hard coded value to check in the checkbox plugin for the DetailView.
I'm currently putting a value on the fake rows I can ignore the rows generated by expanding detailview it. e.g (selectableRow)

**Comment by ghiscoding** [comment]
I'm not exactly sure what problem you are describing, I mean I know you are talking about the Row Detail plugin and perhaps I forgot how it work for that part. But what we have from the `onSelectedRowsChanged` is indeed `(e, args)` which `args` contain `{ grid, rows }` and from there I would typically call 
```js
onGrid1SelectedRowsChanged(e, args) {
  var grid = args.grid;
  args.rows.forEach(function(idx) {
     var item = grid.getDataItem(idx);
  });
```
```

**Comment by SatanEnglish** [comment]
```
var columnpicker = new Slick.Controls.ColumnPicker(columns, grid, options);
columnpicker.canRowBeSelected(function(row, cell, value, columnDef, dataContext) {
  return dataContext.id % 2;
});
```

Ok I guess my issue here then would only be cell and value which you cannot get from the other points were you need disabling code e.g. you cannot do

```
var columnpicker = new Slick.Controls.ColumnPicker(columns, grid, options);
columnpicker.canRowBeSelected(function(row, cell, value, columnDef, dataContext) {
  return cell = 82;
});
```
And expect to be able to stop the click even given some cell number as you cannot get that info at time of the click event.


Guess most practical would be the following as we always have all of them
```
var columnpicker = new Slick.Controls.ColumnPicker(columns, grid, options);
columnpicker.canRowBeSelected(function(row, columnDef, dataContext) {
  
  //(row, columnDef, dataContent)<== all items we should be able to get from all the events we need to   
  //stop
 
  return dataContext.id % 2;
});
```

**Comment by ghiscoding** [comment]
The `cell = 82` is a little extreme since `cell` is actually the column number in the grid, e.g. 3rd column is `cell = 2`.. but actually wait, I tried it and the `cell` here is always `0` since it's the checkbox column, so I could at least remove `cell` from the arguments and also `value` because it's always `undefined`, so the only valid arguments left are: (`row, columnDef, dataContext`) and `grid` was added a year ago as the last argument of a Formatter so the final list of arguments that are valid are: (`row, columnDef, dataContext`, `grid`)

and by the way I already have it working with the override method I talked about 😄 
I changed the name of the function to `selectableOverride` but I can also use `canRowBeSelected`, whichever fits best. The result is the same, see below

```js
  checkboxSelector.selectableOverride(function(row, columnDef, dataContext, grid) {
    return dataContext.id % 2;
  });
```
![selectable_override](https://user-images.githubusercontent.com/643976/53313145-01e3f300-3886-11e9-917e-35180a26e23b.png)

**Comment by SatanEnglish** [comment]
@ghiscoding 
Nice yep if you read my code parts above you can see the functions that need a check are.
`handleSelectedRowsChanged` `handleHeaderClick` and `checkboxSelectionFormatter`

**Comment by SatanEnglish** [comment]
handleSelectedRowsChanged <= This is the one that is call if you hold control and click on a row.
You will need a check here.

**Comment by ghiscoding** [comment]
I look at your code but I'm not sure how to deal with this, on your code it looks like you're adding an extra property on the object, like we do with RowDetail plugin I guess, but I'm not sure when that would be added. The problem that I'm facing now is that I used the Formatter to show or not the checkbox when the row can be selected, but a Formatter uses the viewport cache, it only deals with few rows at a time, so if I keep a `_selectableRows` that is populated through the Formatter (my current code) it only works for what is in the cache... so it's not the best. 

I'm tired and that is enough for tonight, I'm leaving it as it is, anyone can try the code with the PR #346 that I just created, if anyone can find the best way to deal with the "Select All", that is the only problem left to resolve. See the problem below

![irpger3qti](https://user-images.githubusercontent.com/643976/53314604-40c97700-388d-11e9-8287-44401f7d610f.gif)

**Comment by SatanEnglish** [comment]
All good I'll keep looking at it my end.
That's Y I had it on the row data my end that way it's already there.
The data comes from the server side so I already know the data.X to check for the bool property

**Comment by ghiscoding** [comment]
@SatanEnglish 
I think I now understand why you only had 1 argument in the override method, that is `dataContext`. Because that is about the only thing that we have available from the `handleHeaderClick` callback, is that it? I don't think we can get the `columnDef` from there, can we? But I guess the `columnDef` is not necessary anyway since we do the selection by row not by column. If that is the case, then the override method would change into 

```diff
// make only Even Rows selectable by using the override function
- checkboxSelector.selectableOverride(function(row, columnDef, dataContext, grid) {
+ checkboxSelector.selectableOverride(function(dataContext, grid) {
  return dataContext.id % 2;
});
```

**Comment by SatanEnglish** [comment]
Here are your new functions.

I'll let you decide how you want to handle them here is how I've done it.

```
        function checkboxSelectionFormatter(row, cell, value, columnDef, dataContext, grid) {
            var UID = createUID() + row;

            if (dataContext) {
                if (hasCheckBox(row, columnDef, dataContext, grid)) {
                    return _selectedRowsLookup[row]
                        ? "<input id='selector" + UID + "' type='checkbox' checked='checked'><label for='selector" + UID + "'></label>"
                        : "<input id='selector" + UID + "' type='checkbox'><label for='selector" + UID + "'></label>";
                }
                else {
                    return _selectedRowsLookup[row]
                        ? "<input id='selector" + UID + "' type='checkbox' checked='checked'><label for='selector" + UID + "'></label>"
                        : "<input id='selector" + UID + "' type='checkbox'><label for='selector" + UID + "'></label>";
                }
            }
            return null;
        }

        function hasCheckBox(row, columnDef, dataContext, grid) {
            if (typeof _selectableOverride === 'function') {
                return _selectableOverride(row, columnDef, dataContext, grid);
            }
            else {
                return true;
            }
        }

        function handleHeaderClick(e, args) {
            if (args.column.id == _options.columnId && $(e.target).is(":checkbox")) {
                // if editing, try to commit
                if (_grid.getEditorLock().isActive() && !_grid.getEditorLock().commitCurrentEdit()) {
                    e.preventDefault();
                    e.stopImmediatePropagation();
                    return;
                }
                
                if ($(e.target).is(":checked")) {
                    var rows = [];
                    var columnDef = _grid.getColumns(getCheckboxColumnCellIndex);
                    for (var i = 0; i < _grid.getDataLength(); i++) {

                        // Get the row and check it's a selectable row before pushing it onto the stack
                        var rowItem = _grid.getDataItem(i);
                        if (hasCheckBox(i, columnDef, rowItem, _grid)) {
                            rows.push(i);
                        }
                    }
                    _grid.setSelectedRows(rows);
                } else {
                    _grid.setSelectedRows([]);
                }
                e.stopPropagation();
                e.stopImmediatePropagation();
            }
        }

        function toggleRowSelection(row) {

            var columnDef = _grid.getColumns(getCheckboxColumnCellIndex);

            // In case the toggle is done with ctrl or shift check if row is selectable
            var rowItem = _grid.getDataItem(row);
            if (!hasCheckBox(row, columnDef, rowItem, _grid)) {
                return;
            }

            if (_selectedRowsLookup[row]) {
                _grid.setSelectedRows($.grep(_grid.getSelectedRows(), function (n) {
                    return n != row
                }));
            } else {
                _grid.setSelectedRows(_grid.getSelectedRows().concat(row));
            }
            //_grid.setActiveCell(row, getCheckboxColumnCellIndex()); // Commented out on 20/02/2019 as was acting like a double click
            //_grid.focus();
        }

        function handleSelectedRowsChanged(e, args) {
            var selectedRows = _grid.getSelectedRows();
            var columnDef = _grid.getColumns(getCheckboxColumnCellIndex);
            var lookup = {}, row, i;
            for (i = 0; i < selectedRows.length; i++) {
                row = selectedRows[i];

                // If we are allowed to select the row
                var rowItem = _grid.getDataItem(row);
                if (hasCheckBox(i, columnDef, rowItem, _grid)) {
                    lookup[row] = true;
                    if (lookup[row] !== _selectedRowsLookup[row]) {
                        _grid.invalidateRow(row);
                        delete _selectedRowsLookup[row];
                    }
                }
            }
            for (i in _selectedRowsLookup) {
                _grid.invalidateRow(i);
            }
            _selectedRowsLookup = lookup;
            _grid.render();
            _isSelectAllChecked = selectedRows.length && selectedRows.length == _grid.getDataLength();

            if (!_options.hideInColumnTitleRow && !_options.hideSelectAllCheckbox) {
                if (_isSelectAllChecked) {
                    _grid.updateColumnHeader(_options.columnId, "<input id='header-selector" + _selectAll_UID + "' type='checkbox' checked='checked'><label for='header-selector" + _selectAll_UID + "'></label>", _options.toolTip);
                } else {
                    _grid.updateColumnHeader(_options.columnId, "<input id='header-selector" + _selectAll_UID + "' type='checkbox'><label for='header-selector" + _selectAll_UID + "'></label>", _options.toolTip);
                }
            }
            if (!_options.hideInFilterHeaderRow) {
                var selectAllElm = $("#header-filter-selector" + _selectAll_UID);
                selectAllElm.prop("checked", _isSelectAllChecked);
            }
        }
```

**Comment by SatanEnglish** [comment]
@ghiscoding Hope the code above is helpful enough.
I still think simplifying the call to 2 params would be nicer for computation but might not be to much of an over head.

Copy my changes in and have a play anyway they cover everything.
I'd do it on your branch but playing with my own stuff here and copying it into here for you.

**Comment by ghiscoding** [comment]
@SatanEnglish 
I went ahead with my code change specified in my [previous post](https://github.com/6pac/SlickGrid/issues/343#issuecomment-466874323), I only removed `columnDef` out of the equation and got rid of the `_selectableRows` since it's not needed, neither helpful. Instead I use the override method in both the Formatter and inside the `handleHeaderClick` (SelectAll) and it now works as I wanted it. 

What you provided with

**Comment by SatanEnglish** [comment]
You need to update the others that I put in

**Comment by SatanEnglish** [comment]
No I'm not pushing them onto a stack in the formatter!!

**Comment by ghiscoding** [comment]
You put in a lot of code and I don't know which line you modified unless I check line by line, which is too long for me to do now. It's pass midnight and I'm going to bed. You can review my latest commit in the PR #346, I'm out for tonight, need to work tomorrow...

**Comment by SatanEnglish** [comment]
@ghiscoding 
Every change in the above code is the **IF** block that starts with
`                if (checkSelectableOverride(i, columnDef, rowItem, _grid)) {`

I've update the above code to screen shots with highlights so you can find the changes.
All are needed to cover all the cases just put them in and test them

**Comment by SatanEnglish** [comment]
@ghiscoding Ok thanks for the diff heads up now I can post more useful stuff.
Not sure if this or my last updated post of the code is more help but just in case now that I know how to do this here is the code

```diff
        function handleSelectedRowsChanged(e, args) {
            var selectedRows = _grid.getSelectedRows();
            var lookup = {}, row, i;
            for (i = 0; i < selectedRows.length; i++) {
                row = selectedRows[i];

+                // If we are allowed to select the row
+                var rowItem = _grid.getDataItem(row);
+                if (checkSelectableOverride(i, rowItem, _grid)) {
                    lookup[row] = true;
                    if (lookup[row] !== _selectedRowsLookup[row]) {
                        _grid.invalidateRow(row);
                        delete _selectedRowsLookup[row];
                    }
+                }
            }
            for (i in _selectedRowsLookup) {
                _grid.invalidateRow(i);
            }
            _selectedRowsLookup = lookup;
            _grid.render();
            _isSelectAllChecked = selectedRows.length && selectedRows.length == _grid.getDataLength();

            if (!_options.hideInColumnTitleRow && !_options.hideSelectAllCheckbox) {
                if (_isSelectAllChecked) {
                    _grid.updateColumnHeader(_options.columnId, "<input id='header-selector" + _selectAll_UID + "' type='checkbox' checked='checked'><label for='header-selector" + _selectAll_UID + "'></label>", _options.toolTip);
                } else {
                    _grid.updateColumnHeader(_options.columnId, "<input id='header-selector" + _selectAll_UID + "' type='checkbox'><label for='header-selector" + _selectAll_UID + "'></label>", _options.toolTip);
                }
            }
            if (!_options.hideInFilterHeaderRow) {
                var selectAllElm = $("#header-filter-selector" + _selectAll_UID);
                selectAllElm.prop("checked", _isSelectAllChecked);
            }
        }

        function toggleRowSelection(row) {           
            
+            var rowItem = _grid.getDataItem(row);
+            if (!checkSelectableOverride(row, rowItem, _grid)) {
+                return;
+            }

            if (_selectedRowsLookup[row]) {
                _grid.setSelectedRows($.grep(_grid.getSelectedRows(), function (n) {
                    return n != row
                }));
            } else {
                _grid.setSelectedRows(_grid.getSelectedRows().concat(row));
            }
            //_grid.setActiveCell(row, getCheckboxColumnCellIndex()); // Commented out on 20/02/2019 as was acting like a double click
            //_grid.focus();
        }

        function handleHeaderClick(e, args) {
            if (args.column.id == _options.columnId && $(e.target).is(":checkbox")) {
                // if editing, try to commit
                if (_grid.getEditorLock().isActive() && !_grid.getEditorLock().commitCurrentEdit()) {
                    e.preventDefault();
                    e.stopImmediatePropagation();
                    return;
                }
                
                if ($(e.target).is(":checked")) {
                    var rows = [];
                    for (var i = 0; i < _grid.getDataLength(); i++) {

+                        // Get the row and check it's a selectable row before pushing it onto the stack
+                        var rowItem = _grid.getDataItem(i);
+                       if (checkSelectableOverride(i, rowItem, _grid)) {
                            rows.push(i);
+                       }
                    }
                    _grid.setSelectedRows(rows);
                } else {
                    _grid.setSelectedRows([]);
                }
                e.stopPropagation();
                e.stopImmediatePropagation();
            }
        }

        function checkboxSelectionFormatter(row, cell, value, columnDef, dataContext, grid) {
            var UID = createUID() + row;

            if (dataContext) {
+                if (checkSelectableOverride(row, dataContext, grid)) {
+                    return _selectedRowsLookup[row]
+                        ? "<input id='selector" + UID + "' type='checkbox' checked='checked'><label for='selector" + UID + "'></label>"
+                        : "<input id='selector" + UID + "' type='checkbox'><label for='selector" + UID + "'></label>";
+                }
+                else {
                    return _selectedRowsLookup[row]
                        ? "<input id='selector" + UID + "' type='checkbox' checked='checked'><label for='selector" + UID + "'></label>"
                        : "<input id='selector" + UID + "' type='checkbox'><label for='selector" + UID + "'></label>";
+                }
            }
            return null;
        }

+        function checkSelectableOverride(row, dataContext, grid) {
+          if (typeof _selectableOverride === 'function') {
+                return _selectableOverride(row, dataContext, grid);
+            }
+            else {
+                return true;
+            }
+        }
```

**Comment by ghiscoding** [comment]
@SatanEnglish 
I added part of your code, there's 1 place where I cannot use your code that is inside the `checkboxSelectionFormatter`, I need to keep my code because if you only use the new `checkSelectableOverride` that you provide it won't work because it will always go to the `else` and create a checkbox when it shouldn't. So without the 2 cascaded `if`, it doesn't work, so I'm keeping my code as it works.

Only 1 problem left though, when clicking on "Select All", it doesn't change the checkbox to `checked` because it doesn't seem to detect that everything was in fact selected. He seem to care about non-existing selectable. If you click 2x times on it, it stays but that is incorrect behavior

**Comment by SatanEnglish** [comment]
@ghiscoding 
I'll have another play I did a quick rewrite to get it able to post on here I must of missed something. I had a disabled state on my original version.

I'll pull your newest version and find were the issue is.
Might be the 2 lines I have commented out that cause the double click I was seeing earlier

**Comment by SatanEnglish** [comment]
@ghiscoding 
Ah my bad copy and paste issue here is what the formater should of been

```diff
        function checkboxSelectionFormatter(row, cell, value, columnDef, dataContext, grid) {
            var UID = createUID() + row;

            if (dataContext) {
                if (!checkSelectableOverride(row, dataContext, grid)) {
                    return _selectedRowsLookup[row]
                        ? "<input id='selector" + UID + "' type='checkbox' checked='checked' disabled><label for='selector" + UID + "'></label>"
                        : "<input id='selector" + UID + "' type='checkbox' disabled><label for='selector" + UID + "'></label>";
                }
                else {
                    return _selectedRowsLookup[row]
                        ? "<input id='selector" + UID + "' type='checkbox' checked='checked'><label for='selector" + UID + "'></label>"
                        : "<input id='selector" + UID + "' type='checkbox'><label for='selector" + UID + "'></label>";
                }
            }
            return null;
        }
```

**Comment by SatanEnglish** [comment]
@ghiscoding 
Ok so you should be able to add what I have above which shows disabled checkboxes.

Then apparently I just went over board on the checks...

```diff
        function handleHeaderClick(e, args) {
            if (args.column.id == _options.columnId && $(e.target).is(":checkbox")) {
                // if editing, try to commit
                if (_grid.getEditorLock().isActive() && !_grid.getEditorLock().commitCurrentEdit()) {
                    e.preventDefault();
                    e.stopImmediatePropagation();
                    return;
                }
                
                if ($(e.target).is(":checked")) {
                    var rows = [];
                    for (var i = 0; i < _grid.getDataLength(); i++) {

-                        // Get the row and check it's a selectable row before pushing it onto the stack
-                        var rowItem = _grid.getDataItem(i);
-                        if (rowItem.selectableRow !== false) {
                            rows.push(i);
-                        }
                    }
                    _grid.setSelectedRows(rows);
                } else {
                    _grid.setSelectedRows([]);
                }
                e.stopPropagation();
                e.stopImmediatePropagation();
            }
        }
```
This fixes the Select all as all the rows go to the selector but then we ignore the disabled ones so the count for select all is still correct.

Also worth noting I checked this with

```diff
            mainSlick._internal.checkboxSelector.selectableOverride(function (row, dataContext, grid) {
+                return dataContext.id % 2 === 0;
            });
```

True / False result

**Comment by ghiscoding** [comment]
@SatanEnglish 
Thanks for the code update, I'll push that later tonight

> PS: Not sure this is the best way to handle it as the selected rows is now all.
Might have to change it to pass in 2 lists.
Or seperate them inside the handleSelectedRowsChanged would be the other way.
I'll have a quick look at that now to see if I can tidy up what is in the SelectedRows

Not sure what you mean there

**Comment by SatanEnglish** [comment]
@ghiscoding 
Say we have 10 rows odd ones being disabled and we press select all.

_grid.getSelectedRows() <= [0,1,2,3,4,5,6,7,8,9,10]

But should be [0,2,4,6,8,10] 

I'm working on this now.

**Comment by SatanEnglish** [comment]
@ghiscoding 
Ok I don't like it but here is the only way I've manage to get it to work correctly.

We need to put the check back in on `handleHeaderClick` and then brute for check if using the function for disabled ones.


```diff
        function handleHeaderClick(e, args) {
            if (args.column.id == _options.columnId && $(e.target).is(":checkbox")) {
                // if editing, try to commit
                if (_grid.getEditorLock().isActive() && !_grid.getEditorLock().commitCurrentEdit()) {
                    e.preventDefault();
                    e.stopImmediatePropagation();
                    return;
                }
                
                if ($(e.target).is(":checked")) {
                    var rows = [];
                    for (var i = 0; i < _grid.getDataLength(); i++) {

+                        // Get the row and check it's a selectable row before pushing it onto the stack
+                        var rowItem = _grid.getDataItem(i);
+                        if (rowItem.selectableRow !== false) {
                            rows.push(i);
+                       }
                    }
                    _grid.setSelectedRows(rows);
                } else {
                    _grid.setSelectedRows([]);
                }
                e.stopPropagation();
                e.stopImmediatePropagation();
            }
        }

```

Then in the `handleSelectedRowsChanged`


```diff
        function handleSelectedRowsChanged(e, args) {
            var selectedRows = _grid.getSelectedRows();
            var lookup = {}, row, i;
+            var disabledCount = 0;
+            if (typeof _selectableOverride === 'function') {
+                for (k = 0; k < _grid.getDataLength(); k++) {
+                    // If we are allowed to select the row
+                    var dataItem = _grid.getDataItem(k);
+                    if (!checkSelectableOverride(i, dataItem, _grid)) {
+                        disabledCount++;
+                    }
+                }
+            }

            for (i = 0; i < selectedRows.length; i++) {
                row = selectedRows[i];

                // If we are allowed to select the row
                var rowItem = _grid.getDataItem(row);
                if (checkSelectableOverride(i, rowItem, _grid)) {
                    lookup[row] = true;
                    if (lookup[row] !== _selectedRowsLookup[row]) {
                        _grid.invalidateRow(row);
                        delete _selectedRowsLookup[row];
                    }
                }
            }
            for (i in _selectedRowsLookup) {
                _grid.invalidateRow(i);
            }
            _selectedRowsLookup = lookup;
            _grid.render();
+            _isSelectAllChecked = selectedRows.length && selectedRows.length + disabledCount == _grid.getDataLength();

            if (!_options.hideInColumnTitleRow && !_options.hideSelectAllCheckbox) {
                if (_isSelectAllChecked) {
                    _grid.updateColumnHeader(_options.columnId, "<input id='header-selector" + _selectAll_UID + "' type='checkbox' checked='checked'><label for='header-selector" + _selectAll_UID + "'></label>", _options.toolTip);
                } else {
                    _grid.updateColumnHeader(_options.columnId, "<input id='header-selector" + _selectAll_UID + "' type='checkbox'><label for='header-selector" + _selectAll_UID + "'></label>", _options.toolTip);
                }
            }
            if (!_options.hideInFilterHeaderRow) {
                var selectAllElm = $("#header-filter-selector" + _selectAll_UID);
                selectAllElm.prop("checked", _isSelectAllChecked);
            }
        }

```

**Comment by ghiscoding** [comment]
It's ok, I didn't do any of the changes yet since I said I'll do that later tonight. If we do this extra check only when the user override the selectable, then I would be ok with it. 

> PSS: The only case that is not covered here but we can leave for now is when you have a DISABLED checkbox that is TICKED. (Won't end up in selectedRows though don't think this is a sensible case)

I think it shouldn't end up in the selected rows array anyway, so that seems fine to me

**Comment by SatanEnglish** [comment]
@ghiscoding 
Dam it I keep finding weird cases.
Crtl click still highlights the row but doesn't TICK checkbox but adds itself to the selectedList i'll just hunt this one down now... 

PS: added `>=` to the `_isSelectAllChecked` line so if won't affect it for now

**Comment by SatanEnglish** [comment]
@ghiscoding Ok more changes for the weird case were we can still get into `handleSelectedRowsChanged` with disabled items from Ctrl + click.

```diff
        function handleSelectedRowsChanged(e, args) {
            var selectedRows = _grid.getSelectedRows();
            var lookup = {}, row, i;
            var disabledCount = 0;
            if (typeof _selectableOverride === 'function') {
                for (k = 0; k < _grid.getDataLength(); k++) {
                    // If we are allowed to select the row
                    var dataItem = _grid.getDataItem(k);
                    if (!checkSelectableOverride(i, dataItem, _grid)) {
                        disabledCount++;
                    }
                }
            }

+            var removeList = [];
            for (i = 0; i < selectedRows.length; i++) {
                row = selectedRows[i];

                // If we are allowed to select the row
                var rowItem = _grid.getDataItem(row);
                if (checkSelectableOverride(i, rowItem, _grid)) {
                    lookup[row] = true;
                    if (lookup[row] !== _selectedRowsLookup[row]) {
                        _grid.invalidateRow(row);
                        delete _selectedRowsLookup[row];
                    }
                }
+                else {
+                    removeList.push(row);
+                }
            }

            for (i in _selectedRowsLookup) {
                _grid.invalidateRow(i);
            }
            _selectedRowsLookup = lookup;
            _grid.render();
            _isSelectAllChecked = selectedRows.length && selectedRows.length + disabledCount >= _grid.getDataLength();

            if (!_options.hideInColumnTitleRow && !_options.hideSelectAllCheckbox) {
                if (_isSelectAllChecked) {
                    _grid.updateColumnHeader(_options.columnId, "<input id='header-selector" + _selectAll_UID + "' type='checkbox' checked='checked'><label for='header-selector" + _selectAll_UID + "'></label>", _options.toolTip);
                } else {
                    _grid.updateColumnHeader(_options.columnId, "<input id='header-selector" + _selectAll_UID + "' type='checkbox'><label for='header-selector" + _selectAll_UID + "'></label>", _options.toolTip);
                }
            }
            if (!_options.hideInFilterHeaderRow) {
                var selectAllElm = $("#header-filter-selector" + _selectAll_UID);
                selectAllElm.prop("checked", _isSelectAllChecked);
            }

+            // Remove items that shouln't of been selected in the first place (Got here Ctrl + click)
+            if (removeList.length > 0) {
+                for (i = 0; i < removeList.length; i++)  {
+                    var remIdx = selectedRows.indexOf(removeList[i]);
+                    selectedRows.splice(remIdx, 1);
+                }
+                _grid.setSelectedRows(selectedRows);
+            }
        }
```

**Comment by ghiscoding** [comment]
@SatanEnglish 
I believe that I followed all the posts you sent and copied all the necessary, you can double-check the final code. 

There is 1 exception, but please leave that one alone, I still had to put back my code instead of using your code from this [post](https://github.com/6pac/SlickGrid/issues/343#issuecomment-467172008) because again the checkbox every row was still showing checkboxes and they shouldn't. My code works so please forget that one and leave as is since it work, it has 2 `if` but no big deal.

Visually, it seems all fine to me, so good job on that.

**Comment by SatanEnglish** [comment]
@ghiscoding I'll check it again tomorrow  difference now will be I am showing DISABLED checkboxes where are you are removing them.

Removing them will have to do I guess.

**Comment by ghiscoding** [comment]
Ok that works, I added a `onSelectedRowsChanged` to make sure all rows are properly selected, it seems fine too. Is the PR #346 complete then?

**Comment by SatanEnglish** [comment]
@ghiscoding OK TESTED

Still need to comment out the `_grid.focus()` in the `toggleRowSelection` does funcky double clicks in some cases.

Also how about cleaning up a small thing

```
        if (_isSelectAllChecked) {
          _grid.updateColumnHeader(_options.columnId, "<input id='header-selector" + _selectAll_UID + "' type='checkbox' checked='checked'><label for='header-selector" + _selectAll_UID + "'></label>", _options.toolTip);
        } else {
          _grid.updateColumnHeader(_options.columnId, "<input id='header-selector" + _selectAll_UID + "' type='checkbox'><label for='header-selector" + _selectAll_UID + "'></label>", _options.toolTip);
        }
```

Can you put this into a function and call it as currently it's duplicate in 2 places

**Comment by ghiscoding** [comment]
@SatanEnglish 
Ok I did the override method rename to `canRowBeSelected` and added `renderSelectAllCheckbox` to be DRY. Hopefully this is the last change, I'd like to close up on this. And just to be on the safe side, I'd like to do one more test tomorrow with my current project at work and make sure it doesn't break anything.

**Comment by ghiscoding** [comment]
@MrMoronIV 
I merged the PR #346 which addresses your original feature request. You can see a demo of how it works within the PR #346 itself or you can take a look at the [example - checkbox header row](https://github.com/6pac/SlickGrid/blob/master/examples/example-checkbox-header-row.html#L112)

Cheers 😄

**Comment by MrMoronIV** [comment]
@ghiscoding  @SatanEnglish @6pac 

Thank you all, I just confirmed it works! Great addition.
